scalableminds/webknossos
# Periodically check consistency of skeleton tracings

This PR changes the following:

- periodic checks are performed to verify the consistency of skeleton tracings. Since the corrupted tracings we know of are easily recognizable since all edges went missing, I added a simple check which compares the amount of edges to the amount of nodes per tree.
- when sending an error to airbrake, the last 200 redux action types are attached (in the above validation case and in general)
- when an error is thrown, a toast will always be shown. this encourages users to refresh the page (to avoid inconsistent states). also we might get more reports which should help us to lower the overall error rate.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- I did two different tests:
  - I changed the edge collection class to return always a size of 0
  - I removed line 292 (`newTree.nodes.mutableSet(nodeId, activeTree.nodes.get(nodeId));`) in skeletontracing_reducer_helpers to screw up the tracing
- the tests don't mirror the symptoms we experienced perfectly, but I think it's good enough
- to see the error reporting in airbrake, I disabled the localhost filtering in error_handling.js and used prod values for projectId, projectKey and envName (see these sample errors: https://scm.airbrake.io/projects/95438/groups/2399470366835402985?tab=overview and https://scm.airbrake.io/projects/95438/groups/2399463319196723426?tab=overview)
- also, I verified that the toast is not shown when the buckets cannot be loaded for some reason

### Issues:
- contributes to #3632 

------
